RFCs/0004-add-pdk-publish.md

```diff
@@ -0,0 +1,123 @@
+# Add `pdk publish` subcommand
+
+## Summary
+
+Add a new `pdk publish` subcommand that allows publishing a given module tarball to the Forge.  
+*Note that this RFC supersedes [RFC-0003](https://github.com/puppetlabs/pdk-planning/blob/master/RFCs/0003-add-pdk-release.md). While RFC-0003 may eventually still be implemented (likely as part of a PDK 2.x major release), this work represents a smaller scope that is more appropriate to the current PDK 1.x series.*
+
+## Background & Assumptions
+
+- With the addition of module management endpoints to the Forge API and authentication via API keys, [authenticating](https://forgeapi.puppet.com/#section/Authentication) and [publishing](https://forgeapi.puppet.com/#operation/addRelease) to the Forge has a public and stable interface.
+
+- Publishing to the Forge is a natural progression beyond the current `pdk build` subcommand, and is an action that may be performed manually or through an automated process.
+
+## Motivation & Goals
+
+- PDK should provide an easy way for module authors to package and distribute their modules via the Puppet Forge.
+
+- PDK should provide composable commands and actions so that module authors can automate the packaging and distribution of their modules, and integrate those steps into a comprehensive CI/CD pipeline.
+
+
+## Proposed Implementation Design
+
+Implement a new pdk subcommand, `pdk publish`, with the following syntax, options, and arguments.
+
+- Basic syntax:
+    - `pdk publish [--package-dir=<path>] [--force] [tarball.tar.gz]`
+
+
+- Options and arguments:
+    - `--package-dir=<path>`
+      - Use given path as path to find tarball to publish.
+
+    - `--force`
+      - Do not prompt for confirmation.
+
+    - `tarball.tar.gz`
+      - Relative or absolute path to tarball to publish.
+
+- Option combination behaviors:
+    - `(no options/arguments)`
+      - Looks in current module’s `metadata.json` for namespace, module name, and version and then attempts to publish `<namespace>-<module_name>-<version>.tar.gz` from default package path.
```

Comment:
My be best to define more concretely what `default package path` actually is.  e.g. is that working dir? or the directory containing the metadata.json ?

Reply:
Hmm, good point. Working dir seems the most intuitive to me. I think using the resulting `pkg` dir from `pdk build` maybe makes more sense if / when we implement RFC-0003, as the build and push actions would be part of the same `pdk release` subcommand.

What do you think, @scotje?

Reply:
I think it should match where `pdk build` will place the package by default, which I believe is `<moduleroot>/pkg`

where `<moduleroot>` is defined as the location where the module's `metadata.json` is located.

Reply:
So `pdk publish` would assume `pkg` within the working dir is the default package path? Since `pdk build` only works within a directory with a `metadata.json` file, I'm wanting to clarify whether `pdk publish` would use `<working dir>/pkg` vs. determining the location of the `<moduleroot>` in some way.

Reply:
It looks like `pdk build` currently diverges from the behavior of other commands by requiring you to be working in the directory with `metadata.json`.

I think it should be making use of `PDK::Util.module_root` (https://github.com/puppetlabs/pdk/blob/master/lib/pdk/util.rb#L118-L132) so that it can be run from anywhere inside a module tree, and then `pdk publish` can use the same logic to locate the `<moduleroot>`.